### PR TITLE
Implement `get_catalog_for_single_relation` macro

### PR DESCRIPTION
**Pull request by aranke**
### Problem

We need to implement the `get_catalog_for_single_relation` macro and `GetCatalogForSingleRelation` capability for Snowflake.

### Solution

* Set `GetCatalogForSingleRelation` capability to `Full`
* Implement `get_catalog_for_single_relation` macro using `show objects`

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

**Comment by aranke** [comment]
Current `pre-commit` failure:

```
Check for dbt-core usage in an adapter...................................Failed
- hook id: dbt-core-in-adapters-check
- exit code: 1

A dbt-core module is imported in dbt/adapters/snowflake/impl.py: dbt.artifacts.schemas.catalog.TableMetadata
```

**Comment by jtcohen6** [comment]
Nice to see that the `dbt-core` import check is working as intended!

I think that means `TableMetadata, StatsDict, StatsItem` will need to move into `dbt-adapters` &/or `dbt-common`. That might also enable [a better return type](https://github.com/dbt-labs/dbt-adapters/pull/231/files#r1616964911) for the methods introduced in https://github.com/dbt-labs/dbt-adapters/pull/231

**Comment by aranke** [comment]
Tests are currently failing because we haven't merged in https://github.com/dbt-labs/dbt-adapters/pull/231 yet.

**Comment by jtcohen6** [comment]
> Tests are currently failing because we haven't merged in https://github.com/dbt-labs/dbt-adapters/pull/231 yet.

@aranke You can update [this line in dev requirements](https://github.com/dbt-labs/dbt-snowflake/blob/4ff6bdb46b8d1a9282b475931ba753806f1eef0c/dev-requirements.txt#L3) to run full dbt-snowflake CI with your proposed `dbt-adapters` changes:
```
git+https://github.com/dbt-labs/dbt-adapters.git@get_relation_metadata
```

File: dbt/adapters/snowflake/impl.py [hunk 1]
```diff
@@ -129,6 +139,83 @@ def get_columns_in_relation(self, relation):
             else:
                 raise
 
+    def _show_object_metadata(self, relation):
```
Comment (mikealfare) [inline dbt/adapters/snowflake/impl.py:142]: Could we add type annotations?

File: dbt/adapters/snowflake/impl.py [hunk 2]
```diff
@@ -129,6 +139,83 @@ def get_columns_in_relation(self, relation):
...
+                    value=row.get("bytes"),
+                    include=True,
+                    description="Size of the table as reported by Snowflake",
+                ),
+            }
+
+            catalog_columns = {}
+            columns_metadata = self.get_columns_in_relation(relation)
+
+            for i, c in enumerate(columns_metadata):
```
Comment (mikealfare) [inline dbt/adapters/snowflake/impl.py:207]: These lines could be written as a `dict` comprehension. I don't know if it's necessarily better, but for the sake of comparison:
```python
catalog_columns = {
    c.column: ColumnMetadata(type=c.dtype, index=i + 1, name=c.column)
    for i, c in enumerate(self.get_columns_in_relation(relation))
}
```

File: dbt/include/snowflake/macros/adapters.sql [hunk 3]
```diff
@@ -47,6 +47,15 @@
   {% do return(columns) %}
 {% endmacro %}
 
+{% macro snowflake__show_object_metadata(relation) %}
+  {%- set sql -%}
+    show objects like '{{ relation.identifier }}' in {{ relation.include(identifier=False) }} limit 1
```
Comment (colin-rogers-dbt) [inline dbt/include/snowflake/macros/adapters.sql:52]: do we need to use `ilike`?

File: dbt/adapters/snowflake/impl.py [hunk 4]
```diff
@@ -131,6 +140,83 @@ def get_columns_in_relation(self, relation):
...
+        if not object_metadata:
+            return None
+
+        row = object_metadata[0]
+
+        is_dynamic = row.get("is_dynamic")
+        kind = row.get("kind")
+
+        if is_dynamic == "YES" and kind == "BASE TABLE":
+            table_type = "DYNAMIC TABLE"
```
[the remark this answers is not in the document]
Comment (aranke) [reply dbt/adapters/snowflake/impl.py]: @colin-rogers-dbt good call-out, changed to do comparisons against enums.

File: dbt/adapters/snowflake/relation_configs/policies.py [hunk 5]
```diff
@@ -7,6 +7,7 @@
 class SnowflakeRelationType(StrEnum):
     Table = "table"
     View = "view"
+    MaterializedView = "materialized_view"
```
Comment (mikealfare) [inline dbt/adapters/snowflake/relation_configs/policies.py:10]: We don't support MVs for SF. Adding this here suggests that we do.

File: dbt/adapters/snowflake/impl.py [hunk 6]
```diff
@@ -131,6 +141,86 @@ def get_columns_in_relation(self, relation):
...
+
+        if is_dynamic and kind == str(SnowflakeRelationType.Table).upper():
+            table_type = str(SnowflakeRelationType.DynamicTable).upper()
+        else:
+            table_type = kind
+
+        # https://docs.snowflake.com/en/sql-reference/sql/show-views#output
+        is_view = kind in (
+            str(SnowflakeRelationType.View).upper(),
+            str(SnowflakeRelationType.MaterializedView).upper(),
```
Comment (mikealfare) [inline dbt/adapters/snowflake/impl.py:177]: Per my comment on `SnowflakeRelationType`, we don't need to check for `MaterializedView`.

**Review by mikealfare — COMMENTED** [review]
I only noticed the MV thing, otherwise it looks good.

**Review by mikealfare — APPROVED** [review]

**Pull request merged**